### DISCUSSION TITLE
Error with weekOfSeason ?

### DISCUSSION THREAD

**Discussion by greenars-sftwr** [romcal/romcal#384] [Q&A] [answered]
Hi, I have a issue, i think that maybe is on me. 

Testing the dates between 2022-09-26 -> 2022-10-02 the output is 
"weekOfSeason": 26. 

But the week 26 is between 2022-09-25 ->  2022-10-01. 

This means the Week is wrong for 1 Day? Maybe i have a wrong configuration? (I take the week number using several sites like http://www.ibreviary.com )

I am testing this using Romcal 3.0.0 
```
import Romcal from 'romcal';

export default function Tests() {
const romcal = new Romcal({
  scope: 'liturgical',
});

romcal.generateCalendar(2022).then((data1) => {
  console.log("start",data1["2022-09-26"][0]["calendar"]); 
  console.log("end",data1["2022-10-02"][0]["calendar"]); 
});
```

Please let me know if i have a error. 

**In addition of this. The Romcal V2.0, apparently doesn't have this issue.**

**Comment by tukusejssirs** [comment]
Hey, @jctc06! :wave: Thanks for showing up! :wink:

Well, as I understand your issue, the problem lays in whether a week should start on Sunday (25 Sep 2022) or Monday (26 Sep 2022). In the (Roman) Catholic liturgy, a week always starts with Sunday, however, in many parts of the world, it start with Monday.

Week numbers are calculated using [`getWeekNumber()`](https://github.com/romcal/romcal/blob/dev/lib/utils/dates.ts#L53-L81) function. It calculates ISO week numbers (I presume those are used in the RC liturgy). I seems to be alright: I tested the function outside of romcal and it gives correct week numbers and weeks are started with Sunday:

```ts
// 24 Sep 2022
getWeekNumber(new Date(2022, 08, 24))
38

// 25 Sep 2022
getWeekNumber(new Date(2022, 08, 25))
39

// 26 Sep 2022
getWeekNumber(new Date(2022, 08, 26))
39
```

I haven’t got much time now, I’ll try to find some time later today. We will need to:

- reproduce it with romcal output;
- check if start/end of seasons are correctly computed ([here](https://github.com/romcal/romcal/blob/dev/lib/utils/dates.ts#L1216-L1241));
- maybe check other functions too.

/cc @emagnier, I ping you as you created those functions. You could probably find the (potential) issue much more quickly. :wink:

**Reply by greenars-sftwr** [reply]
Hi @tukusejssirs . Thank you for the answer. 

Its quite strange because the The Romcal V2.0 doesn't have the issue, this happen only with the 3.0. Maybe you can try to repoduce it with the date "2022-09-25". The correct "weekOfSeason" should be  "26". 

Thank you again for you time!

**Reply by tukusejssirs** [reply]
> The Romcal V2.0 doesn't have the issue, this happen only with the 3.0.

Just a quick comment, as I still haven’t managed to find some time to delve into this issue: v3 is complete rewrite of romcal made by @emagnier. He fixed quite a lot of stuff, but the code is still quite new and in development, thus issues are to be expected. We also need to create some tests (the coverage is very low ATM).

**Reply by tukusejssirs** [reply]
The following is a calendar for 2022 ~~(note that weeks start with Monday in it)~~. The first column (the one with `W` instead of day of week in header) is the `weekOfSeason`, however, only for the Ordinary time (both before and after Easter). For other season, I left that column empty.

Just some dates in order to help you double-check the weak of season numbers:
- 9 January 2022: The Baptism of the Lord (when the Epiphany of the Lord is on 6 January 2022) or the Epiphany of the Lord;
- 2 March 2022: Ash Wednesday;
- 17 April 2022: Easter Sunday of the Resurrection of the Lord;
- 5 June 2022: Pentecost Sunday.

Note that week 9 of Ordinary time is omitted this year (according to the liturgical rubrics) in order to comply with the rule that the Sunday of Our Lord Jesus Christ, King of the Universe (Christ the King; 34th Sunday) to occur four weeks before the Nativity of the Lord (Christmas).

```c
        January                   February
 W Su Mo Tu We Th Fr Sa    W Su Mo Tu We Th Fr Sa            March
                      1    4        1  2  3  4  5    W Su Mo Tu We Th Fr Sa
    2  3  4  5  6  7  8    5  6  7  8  9 10 11 12    8        1  2  3  4  5
 1  9 10 11 12 13 14 15    6 13 14 15 16 17 18 19       6  7  8  9 10 11 12
 2 16 17 18 19 20 21 22    7 20 21 22 23 24 25 26      13 14 15 16 17 18 19
 3 23 24 25 26 27 28 29    8 27 28                     20 21 22 23 24 25 26
 4 30 31                                               27 28 29 30 31
         April                      May                       June
 W Su Mo Tu We Th Fr Sa    W Su Mo Tu We Th Fr Sa    W Su Mo Tu We Th Fr Sa
                   1  2       1  2  3  4  5  6  7                1  2  3  4
    3  4  5  6  7  8  9       8  9 10 11 12 13 14   10  5  6  7  8  9 10 11
   10 11 12 13 14 15 16      15 16 17 18 19 20 21   11 12 13 14 15 16 17 18
   17 18 19 20 21 22 23      22 23 24 25 26 27 28   12 19 20 21 22 23 24 25
   24 25 26 27 28 29 30      29 30 31               13 26 27 28 29 30

          July                     August                  September
 W Su Mo Tu We Th Fr Sa    W Su Mo Tu We Th Fr Sa    W Su Mo Tu We Th Fr Sa
13                 1  2   18     1  2  3  4  5  6   22              1  2  3
14  3  4  5  6  7  8  9   19  7  8  9 10 11 12 13   23  4  5  6  7  8  9 10
15 10 11 12 13 14 15 16   20 14 15 16 17 18 19 20   24 11 12 13 14 15 16 17
16 17 18 19 20 21 22 23   21 21 22 23 24 25 26 27   25 18 19 20 21 22 23 24
17 24 25 26 27 28 29 30   22 28 29 30 31            26 25 26 27 28 29 30
18 31

        October                   November                  December
 W Su Mo Tu We Th Fr Sa    W Su Mo Tu We Th Fr Sa    W Su Mo Tu We Th Fr Sa
26                    1   31        1  2  3  4  5                   1  2  3
27  2  3  4  5  6  7  8   32  6  7  8  9 10 11 12       4  5  6  7  8  9 10
28  9 10 11 12 13 14 15   33 13 14 15 16 17 18 19      11 12 13 14 15 16 17
29 16 17 18 19 20 21 22   34 20 21 22 23 24 25 26      18 19 20 21 22 23 24
30 23 24 25 26 27 28 29      27 28 29 30               25 26 27 28 29 30 31
31 30 31
```

The same values are in romcal output:

```bash
# Installation
test_folder='r'
mkdir "$test_folder"
cd "$test_folder"
npm i @romcal/calendar.general-roman@3.0.0-dev.11 @romcal/calendar.slovakia@3.0.0-dev.11
```

```js
// Test
const Romcal = require('romcal')
const {Slovakia_En} = require('@romcal/calendar.slovakia')
const r = new Romcal({localizedCalendar: Slovakia_En})
r.generateCalendar(2022).then(i => console.log(i['2022-09-25'][0].calendar.weekOfSeason))

// Output
26
```

Thus I can’t see a problem in here. All week numbers of the Season are correct.

If you still see an issue, post the commands you run in order to help me to reproduce it.

**Reply by emagnier** [reply]
Thank you guys for raising this bug. A bug fix + test coverage is indeed required. Here is the issue to track it: https://github.com/romcal/romcal/issues/386

As @tukusejssirs is pointing out, v3 is a major rewrite of the core library, and there is still a lack of test coverages to make this v3 a final release.

**Reply by tukusejssirs** [reply]
@jctc06, could you test [this test code](https://github.com/romcal/romcal/issues/386#issuecomment-1277222130) please)? If all passes, it should output `${date} : true`, e.g. `2022-11-24 : true`.

Note that you might want to change the locale and particular calendar and (possibly) any other stuff you did when you experienced incorrect `weekOfSeason` values.

Thanks! :pray:

**Reply by tukusejssirs** [reply]
According to my tests, romcal output is as follows (same as we can find on ibreviary.com or lc.kbs.sk):
- 18-24 September 2022 is week 25;
- 25 September – 1 October 2022 is week 26;
- 2-8 October 2022 is week 27;
- 9-15 October 2022 is week 28.

Thus statement that the `weekOfSeason` output by romcal for days between 26 September and 2 October is `26` is incorrect (I proved this by [this test](https://github.com/romcal/romcal/issues/386#issuecomment-1277222130)).

Here’s a simple version of the test based on the code you provided in the description (saved as `test.mjs`, run as `node test.mjs`, `node` version `18.10.0`, `npm` version `8.19.2`). I would provide direct links to the dates on ibreviary.com, however, they save the date config with the PHP session, thus I could not find a way to create a direct link to a specific day (other than today). Therefore I provided links to universalis.com, from where I sourced the `expected` week of season values. Same values are on ibreviary.com and lc.kbs.sk.

```js
import Romcal from 'romcal'
const r = new Romcal({scope: 'liturgical'})

r.generateCalendar(2022).then((data1) => {
  [
    {date: '2022-09-25', expected: 26},  // src: https://universalis.com/20220925/today.htm
    {date: '2022-09-26', expected: 26},  // src: https://universalis.com/20220926/today.htm
    {date: '2022-10-02', expected: 27}   // src: https://universalis.com/20221002/today.htm
  ].forEach(i => {
    const dow = new Date(i.date).getDay()
    console.log(`[${dow}] ${i.date} [e: ${i.expected}] [passed: ${i.expected === data1[i.date][0].calendar.weekOfSeason}] :`, data1[i.date][0].calendar.weekOfSeason)
  })
})

// Output
// [${dayOfWeek}] ${date} [e: ${expectedWeekOfSeason}] [passed: ${whetherTheTestPassed}] : ${weekOfSeasonComputedByRomcal}
[0] 2022-09-25 [e: 26] [passed: true] : 26
[1] 2022-09-26 [e: 26] [passed: true] : 26
[0] 2022-10-02 [e: 27] [passed: true] : 27
```

If I somehow misunderstood your issue or if romcal actually outputs incorrect `weekOfSeason` under some conditions, specify the conditions please. :wink:

**Reply by greenars-sftwr** [reply]
> @jctc06, could you test [this test code](https://github.com/romcal/romcal/issues/386#issuecomment-1277222130) please)? If all passes, it should output `${date} : true`, e.g. `2022-11-24 : true`.
> 
> Note that you might want to change the locale and particular calendar and (possibly) any other stuff you did when you experienced incorrect `weekOfSeason` values.
> 
> Thanks! 🙏

Testing your code @tukusejssirs I had in every week a false value like this. 
`2022-09-25 : false [a:25 === e:26]`

I dont know if the language is the problem, I am testing on TypeScript. 

The syntaxis I used for the require is this. 
`const {Slovakia_En} = require('@romcal/calendar.slovakia/cjs/sk.js')`

I get an error if I test with this
`const {Slovakia_En} = require('@romcal/calendar.slovakia') `

**Reply by tukusejssirs** [reply]
> Testing your code @tukusejssirs I had in every week a false value like this.
> `2022-09-25 : false [a:25 === e:26]`

@jctc06, could you share full MWE? Like a Git repository (at GitLab/GitHub/...) or at JSFiddle or similar?

> I dont know if the language is the problem, I am testing on TypeScript.

TypeScript should not be a problem.

> The syntaxis I used for the require is this.
> `const {Slovakia_En} = require('@romcal/calendar.slovakia/cjs/sk.js')`
>
> I get an error if I test with this
> `const {Slovakia_En} = require('@romcal/calendar.slovakia') `

I think this is expected, as under `calendar.${calendarName}/`, we have `cjs/`, `esm/` and `iife/` folders.

**Reply by greenars-sftwr** [reply]
I just noticed that the day is wrong too in my tests. Maybe this can give any light

This is the full output for 2022-09-26 `console.log(data1["2022-09-26"][0]); `
```
{
    "key": "ordinary_time_26_sunday",
    "date": "2022-09-26",
    "dateDef": {
        "dateFn": "dateOfOrdinaryTime",
        "dateArgs": [
            0,
            26
        ],
        "yearOffset": 0
    },
    "precedence": "UNPRIVILEGED_SUNDAY_6",
    "rank": "SUNDAY",
    "allowSimilarRankItems": false,
    "isHolyDayOfObligation": true,
    "isOptional": false,
    "i18nDef": [
        "seasons:ordinary_time.sunday",
        {
            "week": 26
        }
    ],
    "seasons": [
        "ORDINARY_TIME"
    ],
    "periods": [
        "LATE_ORDINARY_TIME",
        "LATE_ORDINARY_TIME"
    ],
    "colors": [
        "GREEN"
    ],
    "martyrology": [],
    "titles": [],
    "calendar": {
        "weekOfSeason": 26,
        "dayOfSeason": 163,
        "dayOfWeek": 0,
        "nthDayOfWeekInMonth": 4,
        "startOfSeason": "2022-01-11",
        "endOfSeason": "2022-11-27",
        "startOfLiturgicalYear": "2021-11-29",
        "endOfLiturgicalYear": "2022-11-27"
    },
    "cycles": {
        "properCycle": "PROPER_OF_TIME",
        "sundayCycle": "YEAR_C",
        "weekdayCycle": "YEAR_2",
        "psalterWeek": "WEEK_2"
    },
    "fromCalendar": "ProperOfTime",
    "fromExtendedCalendars": []
}
``` 

As you can see, the day is "ordinary_time_26_sunday" and same in the "dayOfWeek": 0 (Sunday).

But September 26, 2022 is a Monday. 

Do you think this may be related to my country location?, or something like that? I test my codes locally.
The weird thing is, the Romcal V2 return the right day.

**Reply by tukusejssirs** [reply]
@jctc06, I think you are right to think that this is a timezone issue. Both @emagnier and me use Central European (Summer) Time (UTC+1 or UTC+2), thus we didn’t think about this issue.

> The weird thing is, the Romcal V2 return the right day.

Well, in `romcal@2` we used `dayjs` package dependency which better supports timezone management and I think that this is why it works there, but does not work in `romcal@3`.

**Comment by tukusejssirs** [comment, marked as answer]
@jctc06, could you test `romcal@3.0.0-dev.20` if it fixes this issue for you? In my local (non-browser) tests, it seems to be fixed.

If it fixes your issue, you could mark this comment as an answer to this discussion. Thanks! :pray:

**Reply by greenars-sftwr** [reply]
Thanks @tukusejssirs, great job!. This dev version fixes the issue. It works correctly!
